tests/pytest.org.md

```diff
@@ -0,0 +1,122 @@
+# Pytest organization proposal
+
+This proposal intends to achieve the following
+  - Have a standard way of categorizing tests
+  - Have some guidelines around test file organization
+  - Have a master wrapper for test execution
+  - Follow common documentation style
+  - Test result collection
+
+## Test categorization
+Leverage pytest custom markers to group tests based on topology, platform, features and connection type.
+
+```
+pytest.ini
+[pytest]
+markers:
+    topology(topo_name): The topologies this particular testcase can run against. topo_name can be individual topology names like 't0', 't1', 'ptf', 'any' or a comma separated like ('t0', 't1') if supported on multiple topologies
+    platform(platform_name): used for platform specific test(broadcom, mellanox, vs etc)
+    feature(feature_name): feature this test is written for. eg. acl, nat
+    connection_type(name): names can be 'fabric' (which indicates the presence of a fanout switch) or 'direct' if a testcase uses directly connected links
+
+```
+conftest.py
+
+```
+def pytest_addoption(parser):
+        parser.addoption("--topology", action="store", metavar="NAME",
+                        help="only run tests matching the topology NAME")
+
+def pytest_runtest_setup(item):
+    toponames = [mark.args for mark in item.iter_markers(name="topology")]
+    if toponames:
+        cfg_topos = item.config.getoption("--topology").split(',')
+        if all(topo not in toponames[0] for topo in cfg_topos):
+            pytest.skip("test requires topology in {!r}".format(toponames))
+    else:
+        if item.config.getoption("--topology"):
+            pytest.skip("test does not match topology")
+
+```
+
+Sample test file: test_topo.py
+
+```
+@pytest.mark.topology('t0', 't1')
+def test_all():
+   assert 1 == 1
+
+@pytest.mark.topology('t0')
+def test_t0():
+   assert 1 == 1
+
+
+@pytest.mark.topology('any')
+def test_any():
+   assert 1 == 1
+
+```
+
+Sample test file: test_notopo.py
+
+```
+def test_notopo():
+   assert 1 == 1
+
+```
+
+Test run
+
+```
+py.test --inventory inv --host-pattern dut1 --module-path ../ansible/library/ --testbed tb --testbed_file tb.csv --topology t1 test_topo.py test_notopo.py -rA
```

Comment:
The topology for testbed is already specified in the testbed file.
Can we retrieve it from there instead of specifying it in the command line? 
And in case test masked by  `topology('any')` don't care about topology and run the test.

Reply:
Agreed. The purpose here is a larger bin of test scenario. At this level, we only distinguish t0/t1/no topology requirement (agnostic)/ptf. This flag is more for organizing test execution.

Reply:
@msosyak, we will be introducing a master wrapper for easier execution of testcases. There we would be deriving the topology from the testbed name and running the tests matching the dut topology if this option isn't specified. The user still can specify this option if they want to run some topology agnostic tests along with the dut topology test.